Feat/#501 Typescript setup & cleanup

ciaranschutte opened the pull request:
- move code into `src` folder
- remove unused dependencies
- remove Flow
- add Typescript with strict settings
  - allowJS is true to incrementally convert to TS project
- update Husky pre-push to typecheck
- add npm scripts to build and watch code

hlminh2000 commented on package.json:
```diff
@@ -3,61 +3,45 @@
+    "clean": "rimraf ./build",
+    "type-check": "tsc -p ./tsconfig.json",
+    "prebuild": "npm run clean",
+    "build": "npm run type-check",
```
is it possible to have a type-check that doesn't run a transpile? also I think the build is gonna have to also copy the static assets (namely the tsv test files) to the build directory too

ciaranschutte replied:
I don’t think nightwatch will run .ts files but I will double check 

Tsv test files are being copied in the post build script using the copy-data-files.js script

hlminh2000 replied:
cool! sorry I missed that :)

ciaranschutte replied:
tried running without a transpile with `ts-node`. Unfortunately nightwatch config still looks for js files only so this wont work

hlminh2000 commented on package.json:
```diff
@@ -3,61 +3,45 @@
     "interactive": "./scripts/interactive.js",
-    "nightwatch": "nightwatch -c ./conf/dev.js",
+    "nightwatch": "nightwatch -c ./build/conf/dev.js",
     "nightwatch-chrome": "nightwatch -c ./conf/dev.js --env CHROME_LATEST",
     "nightwatch-ff": "nightwatch -c ./conf/dev.js --env FIREFOX_LATEST",
     "browserstack": "./scripts/local.runner.js -c ./conf/browserstack-default.js",
     "browserstack-all": "./scripts/local.runner.js -c ./conf/browserstack-chrome.js; ./scripts/local.runner.js -c ./conf/browserstack-firefox.js",
```
need to update these commands to eh?

ciaranschutte replied:
I didn’t address these because I think they’ll change / be removed. Will do some more clean up on these to not leave broken window

ciaranschutte replied:
I didn’t address these because I think they’ll change / be removed. Will do some more clean up on these to not leave broken window

ciaranschutte replied:
I didn’t address these because I think they’ll change / be removed. Will do some more clean up on these to not leave broken window

ciaranschutte commented on package.json:
```diff
@@ -3,61 +3,45 @@
-    "geckodriver": "^1.19.1"
+    "geckodriver": "^1.19.1",
+    "husky": "^4.2.5",
+    "rimraf": "^3.0.2",
```
Using rimraf so it’s platform agnostic 
Runs in prebuild script

ciaranschutte commented on tsconfig.json:
```diff
@@ -0,0 +1,13 @@
+    "noImplicitReturns": true,
+    "noImplicitAny": true,
+    "forceConsistentCasingInFileNames": true,
+    "strict": true,
```
Strict:true covers strictNullCheck
It’s just a convience option for several checks

ciaranschutte commented:
Thanks for feedback! Nice catch on docker file I will update this and do the other tasks I’ve mentioned in comments :)

ciaranschutte commented:
@hlminh2000 
- added some cleanup for npm scripts
- fixed top level `scripts` to use TS modules
- changed npm scripts to build first, so the `dockerfile` looks ok to me without change
- after TS transpiles files loose their `+x` executable permission, so I've added another `postbuild` script to fix this
